File: Javascript/DP/M-931. Minimum Falling Path Sum.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var minFallingPathSum = function (matrix) {
  const n = matrix.length;
  const minPathSum = new Array(2).fill(null).map((_) => new Array(n).fill(0));
  minPathSum[0] = [...matrix[0]];

  for (let i = 1; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const curValue = matrix[i][j];
      minPathSum[i % 2][j] = minPathSum[(i - 1) % 2][j] + curValue;
      if (j - 1 >= 0) {
        minPathSum[i % 2][j] = Math.min(
          minPathSum[i % 2][j],
          minPathSum[(i - 1) % 2][j - 1] + curValue
        );
      }
      if (j < n - 1) {
        minPathSum[i % 2][j] = Math.min(
          minPathSum[i % 2][j],
          minPathSum[(i - 1) % 2][j + 1] + curValue
        );
      }
    }
  }

  return Math.min(...minPathSum[(n - 1) % 2]);
};

const testData = [
  {
    data: [
      [2, 1, 3],
      [6, 5, 4],
      [7, 8, 9],
    ],
    ans: 13,
  },
  {
    data: [
      [-19, 57],
      [-40, -5],
    ],
    ans: -59,
  },
];

for (const data of testData) {
  console.log(data.ans === minFallingPathSum(data.data));
}

File: Python/DP/M-931. Minimum Falling Path Sum.py
class Solution:
    @classmethod
    def minFallingPathSum(self, matrix) -> int:
        n = len(matrix)
        min_path_sum = [[0 for j in range(n)] for i in range(n)]
        min_path_sum[0] = [*matrix[0]]

        for i in range(1, n):
            for j in range(n):
                cur_value = matrix[i][j]
                min_path_sum[i % 2][j] = min_path_sum[(i-1) % 2][j] + cur_value
                if j - 1 >= 0:
                    min_path_sum[i % 2][j] = min(min_path_sum[i % 2][j], min_path_sum[(
                        i-1) % 2][j - 1] + cur_value)
                if j < n - 1:
                    min_path_sum[i % 2][j] = min(min_path_sum[i % 2][j], min_path_sum[(
                        i-1) % 2][j + 1] + cur_value)

        return min(min_path_sum[(n - 1) % 2])


test_data = [
    {
        "data": [
            [2, 1, 3],
            [6, 5, 4],
            [7, 8, 9],
        ],
        "ans": 13,
    },
    {
        "data": [
            [-19, 57],
            [-40, -5]
        ],
        "ans": -59,
    },

]

for data in test_data:
    print(data["ans"] == Solution.minFallingPathSum(data["data"]))
